### ISSUE TITLE
Support for WiFi direct (P2P)

### ISSUE THREAD

**Issue by DavidJRobertson** [espressif/esp-idf#6522]
There have been mentions a few years ago that wifi direct was on the to-do list, but it is still not implemented. 

This is surprising because there are very clear reasons why developers would want to use wifi direct in an IOT project. Specifically, it would greatly improve the user experience of using wifi to connect to the ESP32 device from an Android phone without the need for network infrastructure or provisioning.

Right now the alternative is to either use bluetooth (which is low-bandwidth and has licensing/certification cost issues) or to use soft AP on the ESP32, which provides poor user experience due to the requirement for the smartphone wifi connection to be manually switched, interrupting the phone's normal network connection. At least on some Android devices, it is supported to connect to a normal AP and a wifi direct group at the same time.

I note that the upstream wpa_supplicant project includes support for wifi direct/p2p so I presume this could be made to work on the ESP32.

**Comment by Yugnat4Github** [comment]
Permit me to agree with what is said in this feature request.
If application is mobile and range not short enough for bluetooth, the only way to share data with a device is creating soft AP. That means  users will loose internet connectivity on their devices which is not acceptable... Unless there are other ways I don't know about.
Please consider implementing this.

**Comment by sagb2015** [comment]
@DavidJRobertson - Yes, this is on our TODO list, but it may not be available any time soon since our focus is currently on the protocols that are more important from security perspective (WPA3, PMF, DPP) or that cover more IoT/wireless scenarios (FTM, 11kv, 11r etc). However, depending on the demand we may consider to increase the priority of implementing this feature as well.

**Comment by kaedros** [comment]
We have also been waiting for this feature for many years and it would bring a huge improvement in the experience already described very well by @DavidJRobertson.

We cannot imagine that this demand is not important for everyone who develops for IoT.

**Comment by wagner-rebello** [comment]
This feature can help a lot for IoT usage.
The network configuration using a smartphone app would be very simple as the network do not need to change.
Changing networks on Android or iOS is painful. There are many factors that implies in the network refuses to change.
Using wifi direct "kills" all that factors.

**Comment by mauriciofarina** [comment]
@DavidJRobertson and @sagb2015 
We are also on standby for this feature for a log time!

**Comment by itamardv** [comment]
It would be great

**Comment by jensolsson** [comment]
Need this too. Now we are looking for another chip for our project but the ESP32 would be perfect if it only supported WiFi Direct / P2P

**Comment by TeschRenan** [comment]
I agree that it's not extremely important, but I support inclusion.

**Comment by FlynnMa** [comment]
> Need this too. Now we are looking for another chip for our project but the ESP32 would be perfect if it only supported WiFi Direct / P2P

What is the another chip are you talking about?
...I would be happy to use an alternative if if provides both BT and wifi with wifi-direct.

**Comment by wuujiann** [comment]
I too am looking for this feature.  Hopefully, this feature can be impelmented soon.

**Comment by Torarqui** [comment]
This feature would be decisive in our IoT products! As @wagner-rebello pointed out, developing an App that can connect via Wifi to an device without Wifi-Direct is really hard, and even if you manage to do that, the end result is not user friendly as @DavidJRobertson said. Looking forward to see this implemented in ESP-IDF!

**Comment by writetosumeet** [comment]
Any timeline for Wifi Direct for ESP32?
We are considering using ESP32 for a new product but inability to provide WIFI direct is a big hinderance!

**Comment by maleiter** [comment]
We'd need WiFi Direct too. The lack of WiFi Direct support forces us to use another chip with the next version of our product.

**Comment by jack0c** [comment]
Does anyone try to use DPP, which is more official for IOT device provisioning. There are already some Android phones support this feature, and I think new Android phones is likely to support this feature, just like Android support WPS and P2P.

**Comment by ZantsuRocks** [comment]
> Does anyone try to use DPP, which is more official for IOT device provisioning. There are already some Android phones support this feature, and I think new Android phones is likely to support this feature, just like Android support WPS and P2P.

As iOS don't support this technology and the same for old android phones (the majority in Brazil where our product is sold) i can't even try to lose time testing this.

**Comment by DavidJRobertson** [comment]
> > Does anyone try to use DPP, which is more official for IOT device provisioning. There are already some Android phones support this feature, and I think new Android phones is likely to support this feature, just like Android support WPS and P2P.
> 
> As iOS don't support this technology and the same for old android phones (the majority in Brazil where our product is sold) i can't even try to lose time testing this.

Yeah, nice thing about WiFi direct is that it is backwards compatible with WiFi devices that don't support it 'directly' (just appears as access point, so much the same as existing soft-AP provisioning approaches).

**Comment by adamoskaranikas** [comment]
Hi, are there any updates regarding WiFi Direct support for ESP32 ? Is there a way to maybe adapt ESP-NOW to allow for communication with Wifi-Direct enabled devices other than ESP32's (i.e. ESP32 to Android mobile device). This would be a game changer for us, otherwise we will have to look at other MCU options supporting this feature

**Comment by benjamin-shapiro** [comment]
+1 on this -- current soft-AP implementation with Android is highly undesirable. The ability to directly send a video stream to a mobile device on the ESP32 platform would be transformational for our product.

**Comment by ArnyminerZ** [comment]
Hope this gets added ASAP. I agree that it's not more important than security, but platform-wise I think it would help users and developers quite a lot.

**Comment by AxelLin** [comment]
> Does anyone try to use DPP, which is more official for IOT device provisioning. There are already some Android phones support this feature, and I think new Android phones is likely to support this feature, just like Android support WPS and P2P.

@jack0c 
Just want to clarify that to use DPP the Access Point needs to support WPA3, right?
That is a problem because may users are still using access point without WPA3 support.

**Comment by AxelLin** [comment]
> @DavidJRobertson - Yes, this is on our TODO list, but it may not be available any time soon since our focus is currently on the protocols that are more important from security perspective (WPA3, PMF, DPP) or that cover more IoT/wireless scenarios (FTM, 11kv, 11r etc). However, depending on the demand we may consider to increase the priority of implementing this feature as well.

@sagb2015 
It has been a year since your reply. Is there any update for WiFi direct? (May users are asking this feature).
Can we expect v5.0 to include wifi direct support?

**Comment by romatetemadze** [comment]
Any news on the WIFI-Direct?

**Comment by sagb2015** [comment]
Hi All, 
No we do not have support for WiFi-Direct yet, neither we are considering it for IDF 5.0. However, given the demand we would like to evaluate which chips you would like this features to be on. Given the memory size implication we may target limited subset of the features. WiFi-Direct is a very big feature to support.

**Comment by AxelLin** [comment]
@sagb2015 
Do you have a rough estimated number of the memory that will be taken by WiFi-Direct?

**Comment by jensolsson** [comment]
@sagb2015 any of the esp chips would work for us. 

Kind regards
Jens

**Comment by jensolsson** [comment]
@sagb2015 I started playing with WiFi Direct a while ago, the goal was to make the ESP32 appear as a wifi direct device on an Android phone so I hand crafted a vendor IE with P2P info and used esp_wifi_set_vendor_ie to enable it. I finally got it to work so my esp32 device would show up under WiFi Direct in my Android phone. But it was of course not possible to connect to it yet.

I think the next steps would be to enable WPS and do some kind of p2p handshake in order to determine which of the devices will be master. But as it is possible to set a hint on Android to force it to be either master or not, i do not really know if this handshake is necessary or if it would just work if WPS would be enabled?

Is it really that complicated to get a basic WiFi Direct up and running?

Kind regards
Jens

**Comment by nicklasb** [comment]
This would be very valuable. 
As it stands now, the only implemented and broadly used peer to-peer connectivity is BLE with all its complexities. 
WiFi Direct would enable not only peer-to-peer communication with normal PC:s, and possibly with some other MCU:s and devices. 

I really like ESP-NOW, but it is currently only possible to run on espressif hardware and it seems as if it requires driver-level changes on PC:s which just won't happen on a broader scale. Especially when WiFi Direct exists.

I think that it would be in Espressifs interests to open up this avenue of communication, there are so many use cases where you don't want to have to depend on some wonky WiFi router, hassle with passwords and expose yourself to the very well-developed WiFi hacking world to do some communication. 

(And as BLE is implemented, the cat is sort of out of the vendor-lock-in bag anyway. )

**Comment by andresbrocco** [comment]
+1 Up! Great Feature request! Would be really useful

**Comment by Lucas-Heitele** [comment]
+1

**Comment by johnnyb** [comment]
Would love to have Wifi Direct on an ESP32.  Would have been a game-changer for my current project.  Still might be for the future of the project.  Right now we are having to do a full Linux Single-board computer to get a worthwhile Wifi Direct.

**Comment by nicklasb** [comment]
> Would love to have Wifi Direct on an ESP32. Would have been a game-changer for my current project. Still might be for the future of the project. Right now we are having to do a full Linux Single-board computer to get a worthwhile Wifi Direct.

Exactly! There are a huge number of possible solutions that could benefit from this being implemented.

**Comment by jasta** [comment]
I think the real issue here is that wpa_supplicant would need to be updated, then new APIs added to use the features it exposes.  There appear to be a great many reasons why this would be desirable, such as possibly fixing #10615.  Unfortunately there are a lot of patches on top of esp-idf's fork of wpa_supplicant so unmerging those forks and rebasing on top of the latest will probably not be realistic, at least not without potential breakage.  I wonder if perhaps espressif is open to a strategy in which we experimentally bring in another version of wpa_supplicant and try to just get things building and let the community tinker with the unstable support to collect bug reports.

**Comment by nicklasb** [comment]
> Unfortunately there are a lot of patches on top of esp-idf's fork of wpa_supplicant so unmerging those forks and rebasing on top of the latest will probably not be realistic, at least not without potential breakage. 

If there's already a lot of patches and those patches aren't moving upstream, then i would rather try and be inspired by, rather than try to merge, the upstream code. There is always that point where you have to make that decision.
 
So If I were Espressif, I would rather just try to port it into their fork. The business case would be quite uniquely interoperable hardware. It would probably pays to use those extra resources for maintaining their own supplicant.
Or maybe I am just trying to remind Espressif of the hidden costs of non-reuse. :-)

**Comment by jasta** [comment]
> If there's already a lot of patches and those patches aren't moving upstream, then i would rather try and be inspired by, rather than try to merge, the upstream code. There is always that point where you have to make that decision.

Certainly an option, but still a good amount of work.  Given the age of this p2p support request, I'm guessing espressif isn't going to suddenly prioritize deepening their fork just for p2p support.

> So If I were Espressif, I would rather just try to port it into their fork. The business case would be quite uniquely interoperable hardware. It would probably pays to use those extra resources for maintaining their own supplicant.
Or maybe I am just schrewdly trying to remind Espressif of the hidden costs of non-reuse. :-)

I mean, they're spending a lot of time fixing issues I'd bet money upstream has already fixed, so I think they're probably already aware of the costs: https://github.com/espressif/esp-idf/commits/master/components/wpa_supplicant/src

It's a hard spot for them to be in though, I empathize.  I do think though that forking wpa_supplicant and leaving the ability to merge with upstream was probably the wrong call given how much Wi-Fi is still actively developing by the Wi-Fi Alliance.  They're getting very far behind very fast...

**Comment by Thothore** [comment]
Hello, I would love to have this feature for my project (android auto wireless adapter), the chip I have is a wemos lolin32 v1

**Comment by blaueente** [comment]
Might Wifi Aware / NAN be a suitable replacement?
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_nan.html

**Comment by wagner-rebello** [comment]
> Might Wifi Aware / NAN be a suitable replacement? https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_nan.html

No it's not, many OS do not have this functionality (AKA iOS)

**Comment by blaueente** [comment]
> > Might Wifi Aware / NAN be a suitable replacement? https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_nan.html
> 
> No it's not, many OS do not have this functionality (AKA iOS)

True, but iOS also never actually supported WiFi direct.

**Comment by nicklasb** [comment]
> > > Might Wifi Aware / NAN be a suitable replacement? https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_nan.html
> > 
> > 
> > No it's not, many OS do not have this functionality (AKA iOS)
> 
> True, but iOS also never actually supported WiFi direct.

Hm. TBH, this weakens the case for Wi-Fi Direct significantly. 
The usual Apple-for-purely-Apple-reasons wielding their proprietary lock-in sword as much as they possibly can. 
:-(

**Comment by johnnyb** [comment]
Wifi direct is native compatible with Android.  Thus, making devices that would auto-connect to Android and use IP-based protocols would be greatly simplified with Wifi Direct support.

**Comment by nicklasb** [comment]
> Wifi direct is native compatible with Android. Thus, making devices that would auto-connect to Android and use IP-based protocols would be greatly simplified with Wifi Direct support.

Agreed. And I just realized that I have an Android phone now. :-)
Seriously though, it would be actually be quite helpful. 
Bluetooth is a bit cumbersome and overengineered, so a more straightforward protocol would be nice.

**Comment by johnnyb** [comment]
I think if an ecosystem for Wifi Direct were built up inside enough secondary devices, it would force Apple to either go along with the crowd to participate in the ecosystem, or publish their spec openly.

**Comment by AxelLin** [comment]
> Hi All, No we do not have support for WiFi-Direct yet, neither we are considering it for IDF 5.0. However, given the demand we would like to evaluate which chips you would like this features to be on. Given the memory size implication we may target limited subset of the features. WiFi-Direct is a very big feature to support.

@sagb2015 
It has been a long time, do you have any progress to share?

**Comment by nachiketkukade** [comment]
@AxelLin , we're prioritising some other features and improvements over P2P at this moment (
e.g. - WiFi Aware). So there is no major update on this from our side.

**Comment by artem-pdus** [comment]
Support of wifi-direct would be really great step towards improving user experiences with diy/embedded devices, looking forward to it.

**Comment by manosmic** [comment]
+1

**Comment by Liesegang** [comment]
+1

**Comment by connornishijima** [comment]
+1

**Comment by Akhil-Titus** [comment]
+1

**Comment by voxvanhieu** [comment]
+1

**Comment by AxelLin** [comment]
@nachiketkukade 

This was marked as "Status:Reviewing" since Feb 27, but there is no more update.
How is the status now? (It does not make sense to review for such long time).

**Comment by nachiketkukade** [comment]
> we're prioritising some other features and improvements over P2P at this moment (
e.g. - WiFi Aware). So there is no major update on this from our side.

@AxelLin , there is no change in our roadmap since this update was given. Bot had incorrectly moved this to `Reviewing`.

**Comment by finger563** [comment]
👍 for this feature

**Comment by jesses2025smith** [comment]
+1

**Comment by ozanatwewalkio** [comment]
It looks like it's been a very long time since the first request, any plan to have progress so far?

**Comment by zjzhang-cn** [comment]
+1

**Comment by antonio-deepware** [comment]
+1

**Comment by AHWang2** [comment]
Does ESP support P2P now? Or any other solution for substitution, such as ESP-NOW?

**Comment by nicklasb** [comment]
> Does ESP support P2P now? Or any other solution for substitution, such as ESP-NOW?

ESP-NOW is a proprietary Espressif tech that isn't implemented by any other. By contrast, WIFI-direct is, and would enable interopability with a host of other devices (especially phones in my case) without the odd complexity and low speed of bluetooth.  
I would create a host of interesting use cases.

**Comment by AHWang2** [comment]
Do you think we can implement it on our own, by developers?

**Comment by nicklasb** [comment]
I think that we would need a bunch of proprietary information about the Esp32 wifi implementation. 
But i might be wrong, perhaps there are other implementations one could use as a starting point.

**Comment by AHWang2** [comment]
Yes, I think it's complex. But I create a repository and aim to implement some basic function step by step. I welcome you all to follow this project and give some your proposal. Let's do it. Isn't it cool?

**Comment by AHWang2** [comment]
I recently finish the first step of P2P. I let my android phone find out my ESP32S3 in its P2P list, only by adding some information element to ESP's probe response frame. You can have a look at my project.https://github.com/AHWang2/ESP-P2P-scanner

**Comment by nicklasb** [comment]
Cool, I won't have time for now (until after mew years), but can't help but
thinking that implementing the component would be a good Copilot job if
done stepwise, the standard seems well-defined.

**Comment by AHWang2** [comment]
Good. I'll keep going on this project.

**Comment by AHWang2** [comment]
Hey guys, if WiFi P2P is implemented on ESP, in your application scenario, do you want ESP devices to become GO or Android phones to become GO? And why?

**Comment by jesses2025smith** [comment]
> Hey guys, if WiFi P2P is implemented on ESP, in your application scenario, do you want ESP devices to become GO or Android phones to become GO? And why?

Both

**Comment by AHWang2** [comment]
> > Hey guys, if WiFi P2P is implemented on ESP, in your application scenario, do you want ESP devices to become GO or Android phones to become GO? And why?
> 
> Both

What do you mean "both"? Could you please share your scenario?

**Comment by jesses2025smith** [comment]
ESP devices can as both of GO and client

**Comment by nicklasb** [comment]
> > > Hey guys, if WiFi P2P is implemented on ESP, in your application scenario, do you want ESP devices to become GO or Android phones to become GO? And why?
> > 
> > 
> > Both
> 
> What do you mean "both"? Could you please share your scenario?

If this was implemented, I would usually have the ESP32 as a group owner to be able for a phone to use its services. But eventually I would like mix them to have ESPs using the same services.

So if you are asking which to begin with implementing, it would be GO first, then Client.  At least after my personal priorities. 
But others may of course have other use cases.

(Main key for me is not having to have a router and/or be separate of the wifi)

**Comment by AHWang2** [comment]
> ESP devices can as both of GO and client

Thanks for your replying. May I ask your scenario for ESP development?

**Comment by AHWang2** [comment]
> > > > Hey guys, if WiFi P2P is implemented on ESP, in your application scenario, do you want ESP devices to become GO or Android phones to become GO? And why?
> > > 
> > > 
> > > Both
> > 
> > 
> > What do you mean "both"? Could you please share your scenario?
> 
> If this was implemented, I would usually have the ESP32 as a group owner to be able for a phone to use its services. But eventually I would like mix them to have ESPs using the same services.
> 
> So if you are asking which to begin with implementing, it would be GO first, then Client. At least after my personal priorities. But others may of course have other use cases.
> 
> (Main key for me is not having to have a router and/or be separate of the wifi)

Thanks for your replying. May I ask your scenario for ESP development?

**Comment by nicklasb** [comment]
> > > > > Hey guys, if WiFi P2P is implemented on ESP, in your application scenario, do you want ESP devices to become GO or Android phones to become GO? And why?
> > > > 
> > > > 
> > > > Both
> > > 
> > > 
> > > What do you mean "both"? Could you please share your scenario?
> > 
> > 
> > If this was implemented, I would usually have the ESP32 as a group owner to be able for a phone to use its services. But eventually I would like mix them to have ESPs using the same services.
> > So if you are asking which to begin with implementing, it would be GO first, then Client. At least after my personal priorities. But others may of course have other use cases.
> > (Main key for me is not having to have a router and/or be separate of the wifi)
> 
> Thanks for your replying. May I ask your scenario for ESP development?

I don't have much information I can share currently, I am afraid.

**Comment by jesses2025smith** [comment]
> > ESP devices can as both of GO and client
> 
> Thanks for your replying. May I ask your scenario for ESP development?

Maybe smart home

**Comment by AHWang2** [comment]
> > > ESP devices can as both of GO and client
> > 
> > 
> > Thanks for your replying. May I ask your scenario for ESP development?
> 
> Maybe smart home

I have a doubt here. I think IoT needs low power while WiFi-Direct does cost much power. I was considering if implementing WiFi-Direct based on ESP-IDF was a really valuable requirement. Will there be many ESP developers choosing to use this function? And what's their scenario?

**Comment by jesses2025smith** [comment]
> > > > ESP devices can as both of GO and client
> > > 
> > > 
> > > Thanks for your replying. May I ask your scenario for ESP development?
> > 
> > 
> > Maybe smart home
> 
> I have a doubt here. I think IoT needs low power while WiFi-Direct does cost much power. I was considering if implementing WiFi-Direct based on ESP-IDF was a really valuable requirement. Will there be many ESP developers choosing to use this function? And what's their scenario?

In certain specific scenarios, such as transferring logs from an SD card...

**Comment by Maksons** [comment]
I would like to share possible scenarios/use cases.
It would be some measuring device/test device or sport training device.
Let start with examples:
- Vehicle OBDII test device, connect to car (have enough power) connect to adapter over WiFi direct (from Tablet/Phone/PC). Same time could use Internet to visit webpage which could know how to control particular car and do tests, or send statuses to service center/etc.
- imagine SerialUART / CAN /  I2C / SPI adapter to WiFi direct and programs to operate those online / webpage.
- standalone (not IoT) devices control / update . There are many stand alone medical/sport/etc devices which should not be connected to Internet, but provide big amount of data and application for those need internet (upload/download), or as I say "Application" itself could be webpage.

Take example from existing devices using WiFi direct technology - Printer / Scanner / Photo camera , all of them better do not connect to internet and provide/require big amount of data to user device (Phone/Tablet/PC) which should not loose internet connection.

**Comment by AHWang2** [comment]
> I would like to share possible scenarios/use cases. It would be some measuring device/test device or sport training device. Let start with examples:
> 
> * Vehicle OBDII test device, connect to car (have enough power) connect to adapter over WiFi direct (from Tablet/Phone/PC). Same time could use Internet to visit webpage which could know how to control particular car and do tests, or send statuses to service center/etc.
> * imagine SerialUART / CAN /  I2C / SPI adapter to WiFi direct and programs to operate those online / webpage.
> * standalone (not IoT) devices control / update . There are many stand alone medical/sport/etc devices which should not be connected to Internet, but provide big amount of data and application for those need internet (upload/download), or as I say "Application" itself could be webpage.
> 
> Take example from existing devices using WiFi direct technology - Printer / Scanner / Photo camera , all of them better do not connect to internet and provide/require big amount of data to user device (Phone/Tablet/PC) which should not loose internet connection.

Oh, that's great! I haven't made any contributions to my P2P implementation for some time now. Because I can't find the necessary applicable scenarios. But your words make sense and I will try to put it into practice. Thank you for providing such a specific use case.

**Comment by DavidJRobertson** [comment]
For some applications, [Wi-Fi Aware (NAN)](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#wi-fi-awaretm-nan) may be useful in similar ways to Wifi Direct. An advantage over Wifi Direct is that Wifi Aware is supported on iOS, not just Android. However, I've yet to try implementing it.

**Comment by nachiketkukade** [comment]
> For some applications, [Wi-Fi Aware (NAN)](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#wi-fi-awaretm-nan) may be useful in similar ways to Wifi Direct. An advantage over Wifi Direct is that Wifi Aware is supported on iOS, not just Android

> I think IoT needs low power while WiFi-Direct does cost much power. I was considering if implementing WiFi-Direct based on ESP-IDF was a really valuable requirement. 

@DavidJRobertson , @AHWang2 , agree to the points made above. On ESP-IDF side, Wi-Fi Aware will be a better stack for P2P use case. Apart from its power efficiency and massive push for adoption recently, additional advantages of Wi-Fi Aware are its dynamic Clustering function and a quick Service Discovery mechanism.

It's also a fast evolving protocol, we're upgrading our support to Specs 4.0 to catch up with iOS security requirements (tracked here https://github.com/espressif/esp-idf/issues/16743)
Our current implementation supports Datapath (un-encrypted) and will work with Android phones that support Wi-Fi Aware. Before we catch up to Spec 4.0, we can provide a beta for the P2P use case compatible with Android. It will be upgraded with security over time and eventually support iOS compatibility. Let me know if this will be useful.

**Comment by johnnyb** [comment]
I have not looked into "Wifi Aware".  If people are looking for use cases for Wifi Direct, here is a machine that is actively using it.

https://www.youtube.com/watch?v=ETQHnWBMpiU

ESP chips were considered for this and decided against because of the lack of Wifi Direct support.
